{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import requests\n",
    "import pathlib\n",
    "from requests.auth import HTTPBasicAuth\n",
    "\n",
    "api_key = os.environ.get(\"NCBI_API_KEY\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input\n",
    "taxon_file = \"../data/imported/taxids.list\"\n",
    "\n",
    "# Output\n",
    "output_dir = pathlib.Path(\"../data/generated/download_genomes\")\n",
    "pathlib.Path.mkdir(output_dir, exist_ok=True)\n",
    "genomes_metadata_file = output_dir / \"genomes_metadata.csv\"\n",
    "genome_accessions_file = output_dir / \"genome_accessions.list\"\n",
    "genomes_zip_file = output_dir / \"genomes.zip\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_genome_dataset_report(taxon_id, api_key, params=None):\n",
    "    \"\"\"\n",
    "    Get the genome dataset report for a given taxon id using NCBI API.\n",
    "    Default parameters are:\n",
    "    - reference_only=true\n",
    "    - assembly_source=refseq\n",
    "    - exclude_atypical=true # => don't include genomes that have assembly issues or are otherwise atypical\n",
    "    - assembly_version=current\n",
    "    \"\"\"\n",
    "    if params is None:\n",
    "        params = {\n",
    "            \"filters.reference_only\": \"true\",\n",
    "            \"filters.assembly_source\": \"refseq\",\n",
    "            \"filters.exclude_atypical\": \"true\",\n",
    "            \"filters.assembly_version\": \"current\",\n",
    "        }\n",
    "    auth = HTTPBasicAuth(\"api-key\", api_key)\n",
    "    url = f\"https://api.ncbi.nlm.nih.gov/datasets/v2alpha/genome/taxon/{taxon_id}/dataset_report\"\n",
    "    response = requests.get(url=url, auth=auth, params=params)\n",
    "    if \"reports\" not in response.json():\n",
    "        print(f\"No reports found for taxon id {taxon_id}\")\n",
    "        return\n",
    "    return pd.DataFrame(\n",
    "        _parse_genome_dataset_report_response(response.json(), taxon_id)\n",
    "    )\n",
    "\n",
    "\n",
    "def _parse_genome_dataset_report_response(response, query_taxid):\n",
    "    for report in response[\"reports\"]:\n",
    "        yield {\n",
    "            \"query_taxid\": query_taxid,\n",
    "            \"genome_accession\": report[\"current_accession\"],\n",
    "            \"genome_name\": report[\"organism\"][\"organism_name\"],\n",
    "            \"genome_taxid\": report[\"organism\"][\"tax_id\"],\n",
    "            \"assembly_level\": report[\"assembly_info\"][\"assembly_level\"],\n",
    "            \"assembly_status\": report[\"assembly_info\"][\"assembly_status\"],\n",
    "            \"assembly_name\": report[\"assembly_info\"][\"assembly_name\"],\n",
    "            \"refseq_category\": report[\"assembly_info\"][\"refseq_category\"]\n",
    "            if \"refseq_category\" in report[\"assembly_info\"]\n",
    "            else pd.NA,\n",
    "            \"number_of_contigs\": report[\"assembly_stats\"][\"number_of_contigs\"],\n",
    "            \"contig_n50\": report[\"assembly_stats\"][\"contig_n50\"],\n",
    "            \"contig_l50\": report[\"assembly_stats\"][\"contig_l50\"],\n",
    "            \"genome_coverage\": float(\n",
    "                str(report[\"assembly_stats\"][\"genome_coverage\"]).replace(\"x\", \"\")\n",
    "            )\n",
    "            if \"genome_coverage\" in report[\"assembly_stats\"]\n",
    "            else pd.NA,\n",
    "            \"total_sequence_length\": report[\"assembly_stats\"][\"total_sequence_length\"],\n",
    "            \"total_ungapped_length\": report[\"assembly_stats\"][\"total_ungapped_length\"],\n",
    "        }\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_taxid</th>\n",
       "      <th>genome_accession</th>\n",
       "      <th>genome_name</th>\n",
       "      <th>genome_taxid</th>\n",
       "      <th>assembly_level</th>\n",
       "      <th>assembly_status</th>\n",
       "      <th>assembly_name</th>\n",
       "      <th>refseq_category</th>\n",
       "      <th>model</th>\n",
       "      <th>number_of_contigs</th>\n",
       "      <th>contig_n50</th>\n",
       "      <th>contig_l50</th>\n",
       "      <th>genome_coverage</th>\n",
       "      <th>total_sequence_length</th>\n",
       "      <th>total_ungapped_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>562</td>\n",
       "      <td>GCF_000005845.2</td>\n",
       "      <td>Escherichia coli str. K-12 substr. MG1655</td>\n",
       "      <td>511145</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM584v2</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>1</td>\n",
       "      <td>4641652</td>\n",
       "      <td>1</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>4641652</td>\n",
       "      <td>4641652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>562</td>\n",
       "      <td>GCF_000008865.2</td>\n",
       "      <td>Escherichia coli O157:H7 str. Sakai</td>\n",
       "      <td>386585</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM886v2</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>3</td>\n",
       "      <td>5498578</td>\n",
       "      <td>1</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>5594605</td>\n",
       "      <td>5594605</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  query_taxid genome_accession                                genome_name  \\\n",
       "0         562  GCF_000005845.2  Escherichia coli str. K-12 substr. MG1655   \n",
       "1         562  GCF_000008865.2        Escherichia coli O157:H7 str. Sakai   \n",
       "\n",
       "   genome_taxid   assembly_level assembly_status assembly_name  \\\n",
       "0        511145  Complete Genome         current      ASM584v2   \n",
       "1        386585  Complete Genome         current      ASM886v2   \n",
       "\n",
       "    refseq_category model  number_of_contigs  contig_n50  contig_l50  \\\n",
       "0  reference genome  <NA>                  1     4641652           1   \n",
       "1  reference genome  <NA>                  3     5498578           1   \n",
       "\n",
       "  genome_coverage total_sequence_length total_ungapped_length  \n",
       "0            <NA>               4641652               4641652  \n",
       "1            <NA>               5594605               5594605  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refseq_params = {\n",
    "    \"filters.reference_only\": \"true\",\n",
    "    \"filters.assembly_source\": \"refseq\",\n",
    "    \"filters.exclude_atypical\": \"true\",\n",
    "    \"filters.assembly_version\": \"current\",\n",
    "}\n",
    "\n",
    "genome_dataset_report = get_genome_dataset_report(\"562\", api_key, params=refseq_params)\n",
    "genome_dataset_report\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "141 taxon ids read from ../data/imported/taxids.list\n"
     ]
    }
   ],
   "source": [
    "def read_taxon_list(taxon_file):\n",
    "    \"\"\"\n",
    "    Read a list of taxon ids from a file.\n",
    "    \"\"\"\n",
    "    taxon_list = []\n",
    "    with open(taxon_file) as f:\n",
    "        taxon_list.extend(line.strip() for line in f if not line.startswith(\"#\"))\n",
    "    return list(set(taxon_list))\n",
    "\n",
    "\n",
    "taxon_list = read_taxon_list(taxon_file)\n",
    "print(f\"{len(taxon_list)} taxon ids read from {taxon_file}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No reports found for taxon id 1955243\n",
      "No reports found for taxon id 39491\n",
      "No reports found for taxon id 384638\n",
      "No reports found for taxon id 2053618\n",
      "No reports found for taxon id 1506\n",
      "No reports found for taxon id 29523\n",
      "No reports found for taxon id 84026\n",
      "No reports found for taxon id 1945593\n",
      "No reports found for taxon id 1682\n",
      "No reports found for taxon id 76856\n",
      "No reports found for taxon id 1535\n",
      "No reports found for taxon id 39492\n",
      "No reports found for taxon id 1891238\n",
      "No reports found for taxon id 1869337\n",
      "No reports found for taxon id 1898207\n",
      "No reports found for taxon id 1872530\n",
      "No reports found for taxon id 2023260\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_taxid</th>\n",
       "      <th>genome_accession</th>\n",
       "      <th>genome_name</th>\n",
       "      <th>genome_taxid</th>\n",
       "      <th>assembly_level</th>\n",
       "      <th>assembly_status</th>\n",
       "      <th>assembly_name</th>\n",
       "      <th>refseq_category</th>\n",
       "      <th>model</th>\n",
       "      <th>number_of_contigs</th>\n",
       "      <th>contig_n50</th>\n",
       "      <th>contig_l50</th>\n",
       "      <th>genome_coverage</th>\n",
       "      <th>total_sequence_length</th>\n",
       "      <th>total_ungapped_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>544645</td>\n",
       "      <td>GCF_016889065.1</td>\n",
       "      <td>Butyricimonas virosa</td>\n",
       "      <td>544645</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM1688906v1</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>1</td>\n",
       "      <td>4813143</td>\n",
       "      <td>1</td>\n",
       "      <td>611.9</td>\n",
       "      <td>4813143</td>\n",
       "      <td>4813143</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   query_taxid genome_accession           genome_name  genome_taxid  \\\n",
       "74      544645  GCF_016889065.1  Butyricimonas virosa        544645   \n",
       "\n",
       "     assembly_level assembly_status assembly_name        refseq_category  \\\n",
       "74  Complete Genome         current  ASM1688906v1  representative genome   \n",
       "\n",
       "   model  number_of_contigs  contig_n50  contig_l50 genome_coverage  \\\n",
       "74  <NA>                  1     4813143           1           611.9   \n",
       "\n",
       "   total_sequence_length total_ungapped_length  \n",
       "74               4813143               4813143  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refseq_genomes_dataset_reports = [\n",
    "    get_genome_dataset_report(taxon_id, api_key, params=refseq_params)\n",
    "    for taxon_id in taxon_list\n",
    "]\n",
    "refseq_genomes_df = pd.concat(refseq_genomes_dataset_reports).reset_index(drop=True)\n",
    "refseq_genomes_df.sample()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "representative genome    122\n",
       "reference genome           3\n",
       "Name: refseq_category, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refseq_genomes_df[\"refseq_category\"].value_counts()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Complete Genome    68\n",
       "Scaffold           35\n",
       "Contig             14\n",
       "Chromosome          8\n",
       "Name: assembly_level, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refseq_genomes_df[\"assembly_level\"].value_counts()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17 taxon ids failed to retrieve genome dataset report\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['384638',\n",
       " '2053618',\n",
       " '1869337',\n",
       " '39491',\n",
       " '84026',\n",
       " '1682',\n",
       " '29523',\n",
       " '39492',\n",
       " '1898207',\n",
       " '1872530',\n",
       " '76856',\n",
       " '1891238',\n",
       " '2023260',\n",
       " '1955243',\n",
       " '1535',\n",
       " '1945593',\n",
       " '1506']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "failed_taxon_ids = list(\n",
    "    set(taxon_list).difference(set(refseq_genomes_df[\"query_taxid\"]))\n",
    ")\n",
    "print(f\"{len(failed_taxon_ids)} taxon ids failed to retrieve genome dataset report\")\n",
    "failed_taxon_ids\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_taxid</th>\n",
       "      <th>genome_accession</th>\n",
       "      <th>genome_name</th>\n",
       "      <th>genome_taxid</th>\n",
       "      <th>assembly_level</th>\n",
       "      <th>assembly_status</th>\n",
       "      <th>assembly_name</th>\n",
       "      <th>refseq_category</th>\n",
       "      <th>model</th>\n",
       "      <th>number_of_contigs</th>\n",
       "      <th>contig_n50</th>\n",
       "      <th>contig_l50</th>\n",
       "      <th>genome_coverage</th>\n",
       "      <th>total_sequence_length</th>\n",
       "      <th>total_ungapped_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1924</th>\n",
       "      <td>1955243</td>\n",
       "      <td>GCA_017459635.1</td>\n",
       "      <td>Blautia sp.</td>\n",
       "      <td>1955243</td>\n",
       "      <td>Contig</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM1745963v1</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>109</td>\n",
       "      <td>40166</td>\n",
       "      <td>22</td>\n",
       "      <td>30.0</td>\n",
       "      <td>2669729</td>\n",
       "      <td>2669729</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1095</th>\n",
       "      <td>1898207</td>\n",
       "      <td>GCA_004558435.1</td>\n",
       "      <td>Clostridiales bacterium</td>\n",
       "      <td>1898207</td>\n",
       "      <td>Contig</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM455843v1</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>154</td>\n",
       "      <td>17005</td>\n",
       "      <td>31</td>\n",
       "      <td>32.4</td>\n",
       "      <td>1698269</td>\n",
       "      <td>1698269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1797</th>\n",
       "      <td>1891238</td>\n",
       "      <td>GCA_026417035.1</td>\n",
       "      <td>Burkholderiales bacterium</td>\n",
       "      <td>1891238</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM2641703v1</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>286</td>\n",
       "      <td>42790</td>\n",
       "      <td>27</td>\n",
       "      <td>13.6</td>\n",
       "      <td>3770068</td>\n",
       "      <td>3769698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>878</th>\n",
       "      <td>1898207</td>\n",
       "      <td>GCA_003451875.1</td>\n",
       "      <td>Clostridiales bacterium</td>\n",
       "      <td>1898207</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM345187v1</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>366</td>\n",
       "      <td>5472</td>\n",
       "      <td>82</td>\n",
       "      <td>92.5</td>\n",
       "      <td>1503345</td>\n",
       "      <td>1485905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1620</th>\n",
       "      <td>1891238</td>\n",
       "      <td>GCA_019635025.1</td>\n",
       "      <td>Burkholderiales bacterium</td>\n",
       "      <td>1891238</td>\n",
       "      <td>Contig</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM1963502v1</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>156</td>\n",
       "      <td>30954</td>\n",
       "      <td>26</td>\n",
       "      <td>14.0</td>\n",
       "      <td>2752864</td>\n",
       "      <td>2752864</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     query_taxid genome_accession                genome_name  genome_taxid  \\\n",
       "1924     1955243  GCA_017459635.1                Blautia sp.       1955243   \n",
       "1095     1898207  GCA_004558435.1    Clostridiales bacterium       1898207   \n",
       "1797     1891238  GCA_026417035.1  Burkholderiales bacterium       1891238   \n",
       "878      1898207  GCA_003451875.1    Clostridiales bacterium       1898207   \n",
       "1620     1891238  GCA_019635025.1  Burkholderiales bacterium       1891238   \n",
       "\n",
       "     assembly_level assembly_status assembly_name refseq_category model  \\\n",
       "1924         Contig         current  ASM1745963v1            <NA>  <NA>   \n",
       "1095         Contig         current   ASM455843v1            <NA>  <NA>   \n",
       "1797       Scaffold         current  ASM2641703v1            <NA>  <NA>   \n",
       "878        Scaffold         current   ASM345187v1            <NA>  <NA>   \n",
       "1620         Contig         current  ASM1963502v1            <NA>  <NA>   \n",
       "\n",
       "      number_of_contigs  contig_n50  contig_l50 genome_coverage  \\\n",
       "1924                109       40166          22            30.0   \n",
       "1095                154       17005          31            32.4   \n",
       "1797                286       42790          27            13.6   \n",
       "878                 366        5472          82            92.5   \n",
       "1620                156       30954          26            14.0   \n",
       "\n",
       "     total_sequence_length total_ungapped_length  \n",
       "1924               2669729               2669729  \n",
       "1095               1698269               1698269  \n",
       "1797               3770068               3769698  \n",
       "878                1503345               1485905  \n",
       "1620               2752864               2752864  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genbank_params = {\n",
    "    \"filters.assembly_source\": \"genbank\",\n",
    "    \"filters.exclude_atypical\": \"true\",\n",
    "    \"filters.assembly_version\": \"current\",\n",
    "    \"table_fields\": \"assminfo-accession,assminfo-name\",\n",
    "    \"page_size\": 500,\n",
    "}\n",
    "\n",
    "genbank_genome_dataset_reports = [\n",
    "    get_genome_dataset_report(failed_taxon_id, api_key, params=genbank_params)\n",
    "    for failed_taxon_id in failed_taxon_ids\n",
    "]\n",
    "genbank_genomes_df = pd.concat(genbank_genome_dataset_reports).reset_index(drop=True)\n",
    "genbank_genomes_df.sample(5)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], Name: model, dtype: int64)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genbank_genomes_df[\"assembly_level\"].value_counts()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_taxid</th>\n",
       "      <th>genome_accession</th>\n",
       "      <th>genome_name</th>\n",
       "      <th>genome_taxid</th>\n",
       "      <th>assembly_level</th>\n",
       "      <th>assembly_status</th>\n",
       "      <th>assembly_name</th>\n",
       "      <th>refseq_category</th>\n",
       "      <th>number_of_contigs</th>\n",
       "      <th>contig_n50</th>\n",
       "      <th>contig_l50</th>\n",
       "      <th>genome_coverage</th>\n",
       "      <th>total_sequence_length</th>\n",
       "      <th>total_ungapped_length</th>\n",
       "      <th>gaps</th>\n",
       "      <th>bases_per_contig</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>39491</td>\n",
       "      <td>GCA_022453685.1</td>\n",
       "      <td>[Eubacterium] rectale</td>\n",
       "      <td>39491</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM2245368v1</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>1</td>\n",
       "      <td>3451485</td>\n",
       "      <td>1</td>\n",
       "      <td>129.4</td>\n",
       "      <td>3451485</td>\n",
       "      <td>3451485</td>\n",
       "      <td>0</td>\n",
       "      <td>3451485.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   query_taxid genome_accession            genome_name  genome_taxid  \\\n",
       "4        39491  GCA_022453685.1  [Eubacterium] rectale         39491   \n",
       "\n",
       "    assembly_level assembly_status assembly_name refseq_category  \\\n",
       "4  Complete Genome         current  ASM2245368v1            <NA>   \n",
       "\n",
       "   number_of_contigs  contig_n50  contig_l50  genome_coverage  \\\n",
       "4                  1     3451485           1            129.4   \n",
       "\n",
       "   total_sequence_length  total_ungapped_length  gaps  bases_per_contig  \n",
       "4                3451485                3451485     0         3451485.0  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def select_best_representative(df, consider_coverage=None):\n",
    "    \"\"\"\n",
    "    Select the best representative genome from a genome dataset report.\n",
    "    \"\"\"\n",
    "    if consider_coverage is None:\n",
    "        consider_coverage = True\n",
    "    taxa_dfs = []\n",
    "    # convert columns to correct data types\n",
    "    df = df.astype(\n",
    "        {\n",
    "            \"query_taxid\": \"int\",\n",
    "            \"genome_taxid\": \"int\",\n",
    "            \"number_of_contigs\": \"int\",\n",
    "            \"total_sequence_length\": \"int\",\n",
    "            \"total_ungapped_length\": \"int\",\n",
    "            \"contig_n50\": \"int\",\n",
    "            \"contig_l50\": \"int\",\n",
    "            # \"genome_coverage\": \"float\",\n",
    "        }\n",
    "    )\n",
    "    df[\"gaps\"] = df[\"total_sequence_length\"] - df[\"total_ungapped_length\"]\n",
    "    df[\"bases_per_contig\"] = df[\"total_ungapped_length\"] / df[\"number_of_contigs\"]\n",
    "\n",
    "    if consider_coverage:\n",
    "        # if genome coverage is NA remove it from the dataframe\n",
    "        df = df.dropna(subset=[\"genome_coverage\"]).copy()\n",
    "\n",
    "    # Select the representative genome with the highest genome coverage, least gaps and least contig_l50 and highest contig_n50.\n",
    "    for _, grouped_taxa_df in df.groupby(\"query_taxid\"):\n",
    "        rows = grouped_taxa_df.shape[0]\n",
    "        if rows == 1:\n",
    "            taxa_dfs.append(grouped_taxa_df.iloc[0])\n",
    "            continue\n",
    "\n",
    "        # if a taxon has any 'Complete Genome's then shortlist those genomes to be considered for a representative genome.\n",
    "        if any(grouped_taxa_df[\"assembly_level\"] == \"Complete Genome\"):\n",
    "            grouped_taxa_df = grouped_taxa_df.query(\n",
    "                \"assembly_level == 'Complete Genome'\"\n",
    "            ).copy()\n",
    "\n",
    "        # Select the representative genome with the highest genome coverage, least gaps and least contig_l50 and highest contig_n50.\n",
    "        grouped_taxa_df = grouped_taxa_df.sort_values(\n",
    "            [\"bases_per_contig\", \"contig_n50\", \"contig_l50\", \"genome_coverage\", \"gaps\"],\n",
    "            ascending=[False, False, True, False, True],\n",
    "        )\n",
    "\n",
    "        # append the best representative genome to the taxa_dfs list\n",
    "        taxa_dfs.append(grouped_taxa_df.iloc[0])\n",
    "\n",
    "    return pd.DataFrame(taxa_dfs)\n",
    "\n",
    "\n",
    "select_best_representative(genbank_genomes_df).reset_index(drop=True).sample()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_taxid</th>\n",
       "      <th>genome_accession</th>\n",
       "      <th>genome_name</th>\n",
       "      <th>genome_taxid</th>\n",
       "      <th>assembly_level</th>\n",
       "      <th>assembly_status</th>\n",
       "      <th>assembly_name</th>\n",
       "      <th>refseq_category</th>\n",
       "      <th>number_of_contigs</th>\n",
       "      <th>contig_n50</th>\n",
       "      <th>contig_l50</th>\n",
       "      <th>genome_coverage</th>\n",
       "      <th>total_sequence_length</th>\n",
       "      <th>total_ungapped_length</th>\n",
       "      <th>gaps</th>\n",
       "      <th>bases_per_contig</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>28127</td>\n",
       "      <td>GCF_020735565.1</td>\n",
       "      <td>Hoylesella buccalis</td>\n",
       "      <td>28127</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM2073556v1</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>1</td>\n",
       "      <td>3308411</td>\n",
       "      <td>1</td>\n",
       "      <td>7609.1</td>\n",
       "      <td>3308411</td>\n",
       "      <td>3308411</td>\n",
       "      <td>0</td>\n",
       "      <td>3.308411e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>691816</td>\n",
       "      <td>GCF_000614125.1</td>\n",
       "      <td>Bacteroides rodentium JCM 16496</td>\n",
       "      <td>1236512</td>\n",
       "      <td>Contig</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM61412v1</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>148</td>\n",
       "      <td>78229</td>\n",
       "      <td>19</td>\n",
       "      <td>27.0</td>\n",
       "      <td>4884492</td>\n",
       "      <td>4884492</td>\n",
       "      <td>0</td>\n",
       "      <td>3.300332e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>310297</td>\n",
       "      <td>GCF_902374375.1</td>\n",
       "      <td>Phocaeicola plebeius</td>\n",
       "      <td>310297</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>current</td>\n",
       "      <td>MGYG-HGUT-01364</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>25</td>\n",
       "      <td>478534</td>\n",
       "      <td>4</td>\n",
       "      <td>10.0</td>\n",
       "      <td>4421924</td>\n",
       "      <td>4421324</td>\n",
       "      <td>600</td>\n",
       "      <td>1.768530e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1264</td>\n",
       "      <td>GCF_000179635.2</td>\n",
       "      <td>Ruminococcus albus 7 = DSM 20455</td>\n",
       "      <td>697329</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM17963v2</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>5</td>\n",
       "      <td>3685408</td>\n",
       "      <td>1</td>\n",
       "      <td>30.0</td>\n",
       "      <td>4482087</td>\n",
       "      <td>4482087</td>\n",
       "      <td>0</td>\n",
       "      <td>8.964174e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>40520</td>\n",
       "      <td>GCF_003468995.1</td>\n",
       "      <td>Blautia obeum</td>\n",
       "      <td>40520</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM346899v1</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>35</td>\n",
       "      <td>320905</td>\n",
       "      <td>5</td>\n",
       "      <td>100.0</td>\n",
       "      <td>3889284</td>\n",
       "      <td>3889204</td>\n",
       "      <td>80</td>\n",
       "      <td>1.111201e+05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     query_taxid genome_accession                       genome_name  \\\n",
       "39         28127  GCF_020735565.1               Hoylesella buccalis   \n",
       "120       691816  GCF_000614125.1   Bacteroides rodentium JCM 16496   \n",
       "102       310297  GCF_902374375.1              Phocaeicola plebeius   \n",
       "12          1264  GCF_000179635.2  Ruminococcus albus 7 = DSM 20455   \n",
       "56         40520  GCF_003468995.1                     Blautia obeum   \n",
       "\n",
       "     genome_taxid   assembly_level assembly_status    assembly_name  \\\n",
       "39          28127  Complete Genome         current     ASM2073556v1   \n",
       "120       1236512           Contig         current       ASM61412v1   \n",
       "102        310297         Scaffold         current  MGYG-HGUT-01364   \n",
       "12         697329  Complete Genome         current       ASM17963v2   \n",
       "56          40520         Scaffold         current      ASM346899v1   \n",
       "\n",
       "           refseq_category  number_of_contigs  contig_n50  contig_l50  \\\n",
       "39   representative genome                  1     3308411           1   \n",
       "120  representative genome                148       78229          19   \n",
       "102  representative genome                 25      478534           4   \n",
       "12   representative genome                  5     3685408           1   \n",
       "56   representative genome                 35      320905           5   \n",
       "\n",
       "    genome_coverage  total_sequence_length  total_ungapped_length  gaps  \\\n",
       "39           7609.1                3308411                3308411     0   \n",
       "120            27.0                4884492                4884492     0   \n",
       "102            10.0                4421924                4421324   600   \n",
       "12             30.0                4482087                4482087     0   \n",
       "56            100.0                3889284                3889204    80   \n",
       "\n",
       "     bases_per_contig  \n",
       "39       3.308411e+06  \n",
       "120      3.300332e+04  \n",
       "102      1.768530e+05  \n",
       "12       8.964174e+05  \n",
       "56       1.111201e+05  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genomes_df = pd.concat(\n",
    "    [\n",
    "        select_best_representative(refseq_genomes_df, consider_coverage=False),\n",
    "        select_best_representative(genbank_genomes_df),\n",
    "    ]\n",
    ").reset_index(drop=True)\n",
    "genomes_df.to_csv(genomes_metadata_file, index=False)\n",
    "genomes_df.sample(5)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Check to make sure there are no duplicate genome accessions\n",
    "assert all(\n",
    "    genomes_df[\"genome_accession\"].value_counts() == 1\n",
    "), \"Duplicate genome accessions found!\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the genome accessions to a file\n",
    "genomes_df[\"genome_accession\"].to_csv(genome_accessions_file, index=False, header=False)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download the genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$api_key\" \"$genome_accessions_file\" \"$genomes_zip_file\"\n",
    "\n",
    "datasets download genome accession \\\n",
    "    --api-key $1 \\\n",
    "    --inputfile $2 \\\n",
    "    --no-progressbar \\\n",
    "    --filename $3\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "trees",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
